### ISSUE TITLE
Rewriting front end as a Single Page Application

### ISSUE THREAD

**Issue by mtakac2** [exercism/exercism.io#2296]
### Motivation
The goal is to create completely separate front end that will communicate with the back end using JSON. The main application `exercism.io` would therefore serve just as an API. Front end would be responsible for routing, rendering templates etc. Main motivation is:

* Client side will be independent of server side application. Front end developers will be able to develop new features without having to go through ruby code (and vice versa).
* Better scalability - we will be able to run client on different server and easily connect it to all other services we could need in the future.
* Application will be much faster: client is loading only data it needs - no full page reloads, no unnecessary client - server trips.
* Splitting concerns - Both client and server applications will be easier to extend and maintain.

###  Process 
We have two different ways towards our objective:

1. *Start building new front end as a part of the current application* and incrementally add existing features, separate it from the back end only after we are finished.
2. *Start with the front end as a separate application from the beginning* and
rewrite all the existing (front end) functionality before hooking it up with current application.

Both approaches have their's pros and cons, but in this case, I suggest starting with the separate front end app from the scratch. We can still use most of the existing code (all styles, almost all of HTML, and luckily there currently isn't so much Javascript code). 

I think we should create one long running branch and merge it back after we are done. This way we won't have to deal with maintaining both new and old front end at the same time.

A downside is that we will need to change quite a bit of the back end as well. 
As I mentioned before, routing and templates will be on the client and we will need back end to serve JSON instead of HTML. Simple solution could be if we just change `.erb` templates to some objects generating JSON. Also if we want to keep current URLs unchanged we will need to move the back end endpoints under some namespace / sub-domain (at least if we want to run front end and back end on the same server).

#### Tools
* *AngularJS* - that's probably clear - we are already using it and form my point of view only other option is React, which seems to be really cool, but I am lacking experience for writing project like this in it. 
* We should aim for very modular application, and here comes to play *Webpack*. Really great tool that will allow us bundle the modules together. The best part is that we will be able to always load just the code we really need.
* Next, we will need some task runner. My personal preference is *Gulp*, other option would be Grunt, but let's just say that Gulp is something like Grunt on steroids :).
* And last but not least: *Bower * for dependency management

#### Implementation
* Modular, split by functionality (submissions, nitpicks, teams etc.) and some re-usable global parts (nav, sidebars etc.)
* Would be great if we could get rid of jQuery and use only Angular.
* Coffeescript vs. Javascript: cleaner code  vs lower entry level for potential contributors
* SASS
* HTML or some other templating language? Basically same probelm as Coffescript vs. Javascript
* Styleguide: Especially if we'll go with  Javascript we need clear coding style, otherwise it will be just a mess in few months (AirBnB Js Style Guide, JS Hint, ES Hit)
* Same goes for styles
* Tests: Jasmine, Karma, Protractor
* Overall I would try to keep it down with plugins (meaning use only well written, well maintained and only if they bring some real value). Here are some basic plugins which I find really pleasant to work with: ui-router, ng-template, ui-bootstrap, maybe ng-resource, plus of course some official angular libraries.

### Next steps
1. Discuss details and come to agreement upon basics
2. Come up with workflow (branch, separate front end team etc.)
3. Split the application into logical parts - modules
4. Create basic setup (something similar to template mentioned by @KaoruDev)
5. Start working on individual modules
  1. move corresponding back end endpoints under the namespace
  2. rewrite the end point so it can respond with JSON
  3. move HTML from relevant views into client app
  4. write angular code
6. Connect our shiny new front end with the back end :)

Sorry for long issue - I just wanted to put everything in one place for future reference. There are obviously many moving parts and any comments, suggestions or ideas are hugely appreciated.

**Comment by kytrinyx** [comment]
This is an excellent writeup.

> We have two different ways towards our objective:
> 1. _Start building new front end as a part of the current application_ [...]
> 2. _Start with the front end as a separate application from the beginning_ [...]

Based on my experience, I lean very, very heavily towards (1).

I have already had experience trying to do full or partial rewrites, and I feel like the risks are much higher, even though one could potentially move much more quickly. Much work by people I care about has languished and eventually gone nowhere.

The more I experiment with careful partial rewrites, the more of a fan I become of them.

It's more work because we have to take very small, careful steps, but the wins come much more quickly. I have a dread of long-lived branches.

I would really like to explore what it would take to do a step-by-step cleanup that gets pulled into master and deployed on a regular basis.

**Comment by kytrinyx** [comment]
Oh, and a couple of other random thoughts:

> if we want to keep current URLs unchanged

No, no reason to do that.

> Coffeescript vs. Javascript: cleaner code vs lower entry level for potential contributors

Javascript. I care more about being accessible to more contributors than about a ruby-flavoured and ruby-centric variant that some people might like more.

> Gulp vs Grunt

Gulp is great. Happy to use that.

**Comment by mtakac** [comment]
> Based on my experience, I lean very, very heavily towards (1).

Yeah, when I think about it I must admit I have very similar experience. Although I would still argue that in this case it the rewrite shouldn't be problem, because the application isn't so javascript heavy. Anyway let's play it safe. I will create basic setup and docs for the new client, we can add it next to the existing one, use both and slowly move stuff from back end to new client. 

I guess, the easiest thing is to start with dashboard. That means I need to change the endpoint in sinatra so it responds with JSON, move HTML and CSS into coresponding front end module and write needed angular code. From there we can move on to submissions, teams, nitpicks etc.

Let me know if this works for you and I start working on the first PR - setup of the client.

**Comment by kytrinyx** [comment]
OK, I've been mulling this over for a couple of hours.

Here are some of the things that I think are important to bring into this discussion:

Exercism is still in many ways a prototype. We're working on figuring out how to improve the most important part of the experience: the conversations. In order to do this, we need to focus on figuring out how to encourage people to participate in those conversations (asking questions, critiquing solutions), and how to improve the quality of the critiques. Part of doing this is improving the onboarding experience so that people understand what exercism is about, how it works, and how to get started.

Getting the front-end under control is extremely important, since it seems like it's kind of impossible to make any sort of progress in the application without it.

Everything else, while important, is secondary.

Before you start setting up the client, can we talk a bit about what the technical/build/deploy/etc requirements will be for it?

**Comment by kytrinyx** [comment]
One more thing: I know that there are blind developers using exercism. I want to make sure that from the very beginning we ensure that they can continue to use the core features of the application (we can use lynx to test, they've told me that if it works with lynx it tends to work with their screen readers).

**Comment by kytrinyx** [comment]
> let's play it safe.

Yeah, I think that makes good sense.

> I guess, the easiest thing is to start with dashboard.

Do you mean the page you land on when you log in?

> That means I need to change the endpoint in sinatra so it responds with JSON.

Let's talk about the APIs that are necessary. I'll mostly want to write brand new ones.

> move HTML and CSS into coresponding front end module and write needed angular code. From there we can move on to submissions, teams, nitpicks etc.

I wonder if this would be a good opportunity to start doing some new design/thinking to replace and improve each section instead of just porting it.

**Comment by KaoruDev** [comment]
> Would be great if we could get rid of jQuery and use only Angular.
I'm more familiar with Backbone / jQuery than with Angular so I have to ask, does Angular have all the things that jQuery provide? ( Ajax, Deferred, Object manipulation, etc? ) if yes that is awesome :+1: 

*Nitpick on SPA*
Some pages will have a better flow without a hard refresh, i.e: submitting comments. My biggest concern with SPAs is the danger of memory leaks and deep url links. Again I'm not familiar with Angular so if the framework takes care most of this then :+1: 

@mtakac Do you have experience setting up protractor? I'd love for the build to run tests before deploying to S3. Or even better be able to run that on a CI. We can also add linters to our gulp deploy so that builds will fail when the js code is not inline with our styles.

**Comment by mtakac** [comment]
@kytrinyx 
> Exercism is still in many ways a prototype. We're working on figuring out how to improve the most important part of the experience: the conversations.

I guess important part of this is notifications stream. Apart from that, I don't have any specific ideas now. Do you have some concrete changes on your mind?

> Before you start setting up the client, can we talk a bit about what the technical/build/deploy/etc requirements will be for it?

Sure, I just thought it will be easier to discuss implementation details in individual PRs. From my point of view it will be pretty standard application. Maybe I can post here proposed structure and explain how will things fit together there?

> The more I think about it, the more I'd like to discuss the trade-offs between SPA and a more integrated approach, at least in the medium term.

I am not sure I fully understand this one. What do you mean by more integrated approach?

> I know that there are blind developers using exercism.

I guess this is more a thing of accessibility, right? I am not really expert here, but I suppose this is closely connected with HTML and some of it's attributes. We should make it part of the front end style guide.

> Do you mean the page you land on when you log in?

Yes, that one.

> Let's talk about the APIs that are necessary. I'll mostly want to write brand new ones.

Ok, cool. I'll try to make a list of most of the API endpoints we'll need.

> I wonder if this would be a good opportunity to start doing some new design/thinking to replace and improve each section instead of just porting it.

Hmm, I am not really sure about this. The thing is that we should have a clear goal. If we are going to make some bigger changes and improvements together with porting of the app, chances are we are never going to finish and we'll be stuck with half of the app on the client and the other half on the server. 

@KaoruDev 
> ... does Angular have all the things that jQuery provide?

Yes, jQuery is often use together with angular for DOM manipulation, but that's really bad practice. Working with things, like HTTP requests, deferred objects, etc. is IMO far more pleasant in Angular.

> My biggest concern with SPAs is the danger of memory leaks and deep url links. 

Well of course memory leaks are possibility but they are not so easily caused in Angular as they were in Bakbone. We can minimize risk of bugs like this with good architecture (isolating modules), and few simple tricks.

> Do you have experience setting up protractor?

Yes, we will definitely make protractor (and unit tests) tests part of our CI.

**Comment by KaoruDev** [comment]
cool beans, if you know of any resources I should look into to speed my Angular skills please let me know =)

**Comment by kytrinyx** [comment]
>  From my point of view it will be pretty standard application. Maybe I can post here proposed structure and explain how will things fit together there?

Yeah, I realize that my biggest concern at the moment is actually ops/deployment. I don't have any resources to handle that at the moment. Almost all of exercism lives on heroku, and I'm not sure what's involved in making an SPA live.

> What do you mean by more integrated approach?

Sorry - I meant a server-side approach where angular is used to introduce mini-SPAs, but where the server is serving the templates still.

That would let us swap out all the front-end infrastructure without introducing new requirements in terms of deployement. That said, if we can keep the same deployment/hosting going forward even with SPA, then that seems cleaner.

Eventually I want to move everything to rackspace, but (like I said) time/resources is a problem at the moment.

> Hmm, I am not really sure about [whether this would be a good opportunity to start doing some new design/thinking to replace and improve each section instead of just porting it]. [W]e should have a clear goal. If we are going to make some bigger changes and improvements together with porting of the app, chances are we are never going to finish and we'll be stuck with half of the app on the client and the other half on the server.

Yepp. Good point. I'm convinced :)

It sounds like the accessibility concern can be addressed, and the remaining concern that I have is about the deployment/ops resources. I suspect that this will also be easy to address, I just don't have enough information about it yet.

**Comment by KaoruDev** [comment]
>Sorry - I meant a server-side approach where angular is used to introduce mini-SPAs, but where the server is serving the templates still.

>That would let us swap out all the front-end infrastructure without introducing new requirements in terms of deployement. That said, if we can keep the same deployment/hosting going forward even with SPA, then that seems cleaner.

So it sounds like we first should first rip out lineman and swap it with a gulp task runner that will compile the existing code. Then we can chip away at refactoring?

**Comment by kytrinyx** [comment]
That's a possibility, but it would probably be cleaner to leave everything the way it is and go with @mtakac's plan that creates the seed of the new client completely separately, and then slowly switches piece by piece over.

**Comment by KaoruDev** [comment]
I guess I'm not fully understanding how the js / stylesheets are going to be deployed. Are they being checked into this main repo or are they going to live on S3?

**Comment by mtakac** [comment]
> Yeah, I realize that my biggest concern at the moment is actually ops/deployment. I don't have any resources to handle that at the moment. Almost all of exercism lives on heroku, and I'm not sure what's involved in making an SPA live.

No need to complicate things here at the moment. It will work more less in same way like now. Webpack will take care of creating a bundle file and only thing you need to do is include this file into application. Our entry point can still be served from sinatra, so no change here.

> So it sounds like we first should first rip out lineman and swap it with a gulp task runner that will compile the existing code. Then we can chip away at refactoring?

I would rather just create new client that would be loaded together with old one and we can slowly pull things from the old one to the new one. It's actually not so much code, so it's not gonna be an issue.

> I guess I'm not fully understanding how the js / stylesheets are going to be deployed. Are they being checked into this main repo or are they going to live on S3?

Same way as they are now, we just moved them under the new client and simple build task will bundle and copy them so they can be deployed without any change.
@kytrinyx Are we using S3 now?

I am working on document explaining more the structure, development, build and everything else regarding client and ilustrating the main concepts with some examples. I will link it in a while - Just to make sure we are on the same page.
@KaoruDev I will also include some useful resources about Angular and other tools we mentioned for you. Meanwhile, you can check out the official (angular docs)[https://docs.angularjs.org/api] and (tutorial)[https://docs.angularjs.org/tutorial] if you haven't already.

**Comment by KaoruDev** [comment]
ah ok that makes sense thanks for clearing that up, looking forward to the docs :+1:

**Comment by kytrinyx** [comment]
@mtakac This sounds excellent. We're on heroku at the moment, but we should be able to use their buildpack so that we don't have to commit the generated css/js.

**Comment by mtakac** [comment]
@kytrinyx hmm I am not familiar with heroku buildpack. Does it mean that it will run our build task during the deployment?

**Comment by mtakac** [comment]
@kytrinyx how does it work now? Do you upload contents of the `lib/app/public` folder somewhere or are those generated during the deployment?

**Comment by kytrinyx** [comment]
@mtakac For a while it worked great--when the code shipped to heroku, the buildpack would download dependencies, build the necessary files, and then delete itself (or something). Then I had a snafu with a buildpack for lineman going out of date and I gave up debugging it. The current process is (embarrassingly) that I just generate the thing, add a commit and push it up.

**Comment by mtakac** [comment]
Okay so I just made a PR (#2308) with first version of docs - It's nowhere near complete but it should help to get us started.

@kytrinyx Deployment then won't be an issue. I noticed exercism is using Travis for CI - that's great - should be really easy to set things up there. We can even make it, so that build would failed for code that doesn't meet the minimal standards we set. I know that CI and deployment can be pain. Check out https://codeship.com/ - it's dead simple and it even creates a staging enviroment for you. (Not saying there's anything wrong with travis - just wanted to point out some other options)

**Comment by kytrinyx** [comment]
Awesome -- yeah, let's definitely set up Travis CI to fail for code that fails to meet the standards (only in the new sections of the codebase, of course). I've done this for one of the ruby repos, and it's really great to have early warning for that. We can look into other options, too, of course, though it's not high on my priority list if we can work with what we have at the moment.

**Comment by IanVS** [comment]
It sounds like Angular.js has been decided on, but I just want to throw it out there that it is possible to refactor into Ember.  It doesn't have to be a wholesale transition.  I don't have much experience with it myself, but there is a good podcast about it here: https://frontsidethepodcast.simplecast.fm/3 and a very good (and hilarious) RailsConf video as well: (https://www.youtube.com/watch?t=865&v=VMmaKj8hCR4).  

Ember has a lot of exciting things going on, like FastBoot (server-side rendering), Glimmer, and ember-cli.  Might be worth at least taking a look at.

**Comment by kytrinyx** [comment]
> it is possible to refactor into Ember

I don't have a lot of experience with it, but from the conversations I've had with it, it does not seem like Ember is a good fit for gradual transitions. I agree that Ember has lots of great stuff going on with it, but with exercism where it is right now it doesn't really matter what technology we choose. It needs to be something that is common enough that people can contribute to it, and it needs to solve the basic technical need.

We don't have scaling issues (we're not doing any marketing, we're not trying to bring a million people to the site), we're still in the phase of trying to figure out the core functionality.

**Comment by ErikSchierboom** [comment]
I would like to argue strongly *against* CoffeeScript. It's syntax is something most people are not familiar with, so using plain JavaScript seems like a much better option to me. This would also pave the way to using TypeScript, as that is the language of choice for Angular 2.0.

**Comment by kytrinyx** [comment]
> I would like to argue strongly against CoffeeScript.

Yes, that's my argument above.

**Comment by purezen** [comment]
Nice! Had been looking forward to this.
Though, I would still like to ask how Backbone (or any similar library) would do the job here? Also, wouldn't adding a framework increase the contribution bar?

**Comment by mtakac** [comment]
Regarding the questions about SPA JS frameworks:

There are lots of choices here. We could use Backbone, Ember, React (probably my favorite at the moment) and plenty other FWs just as well as Angular. They all have their pros and cons and to a point, it will always be question of personal experience / preference. BUT I still think there are some strong objective reasons why to choose Angular for this particular project:

1. **Popularity**: with 1000+ contributors, almost 3500 watchers and 40 000 starts it's clearly the most popular JS framework at the moment. That means more possible contributors, more libraries and plugins and more high quality online resources.
2. **Stability**: Backed by major company with significant funds for further development, clear plan for future and many production applications using it's quite safe bet for a long run.
3. **Low entry barrier**: Maybe apart from React, I really don't know any other JS framework with so steep learning curve. You can do relatively a lot with some basic JS knowledge and understanding few fundamentals Angular principles.
4. **Modularity**: Angular with it's dependency injection system really promotes modularity and together with some modern JS dev tools, we can achieve quite impressive level of isolation. Not only that ruby developers won't have to deal with JS, same goes for somebody who wants just to change styles a bit or add some HTML. Maybe even more important is that thanks to this isolation of modules we won't need to deal with conflicts among them.

Of course we will first need to have good architecture in place, well documented code and contribution guides and contribution bar will be higher than if we would use vanilla JS or jQuery, but it should still be quite easy for people to hop in and contribute. 

BTW Making contributions easy is nice but I think that main goal here is to make the application better for the users. From this POV, SPA seems to be the right choice and for that we definitely need a framework - trying to implement something similar in plain JS would lead to such a mess that contributions would be much harder.

**Comment by bIgBV** [comment]
@mtakac I agree with all the points, but I do have a small question though. Wouldn't the client have to be written all over again in angular 2.x because angular 1.x support will be dropped an year or two from now ?

**Comment by mtakac** [comment]
@bIgBV Good question! I thought about it a lot when 2.0 was announced. To be honest I even got to the point where I thought that I won't use Angular for any new projects. Anyway when I got the facts straight I realized it's really not such a big deal as it looks:

1. Angular 2.0 will be released (soonest) in late 2015, Angular team announced support for 1.3 for 18 - 24 months after 2.0 release. That means support will be dropped in late 2017. worst case.
2. Angular team announced there will be upgrade path after all. Sure, it will be probably much more work than just upgrade between minor versions, but considering the time Angular will save us by then, I would say it's pretty much worth it.
3. Even after official support is dropped: Angular is open source and I would bet my money that there will be enough contributors to keep 1.X running for quite some time.
4. There are many things, considered best practice, we can do today, that will help us with eventual migration (architecture should evolve around components, not controllers, using of view models instead of scopes etc.)
5. Every project that "lives" and has any future needs not only to constantly add new features but also to constantly improve it's code base - who can tell what technologies will be used in 3 - 4 years in exercism?

**Comment by bIgBV** [comment]
@mtakac Fair points, all of them. 

Is there any particular component that you would suggest a beginner at angular to build ? I want to contribute to the frontend but I do not want to bite more than I can chew right now.

**Comment by mtakac** [comment]
@bIgBV I would suggest to get familiar with writing directives and directives unit tests.

For example:
**Like box** (see `/submissions/:id`)

Take an array of users who already liked the submission and methods for like / unlike submission. Display the users and implement toggling.
```
<ex-like-box users="[{...}]" on-like="likeSubmission(id, user)" on-dislike="dislikeSubmission(id, user)"></ex-like-box>
```


**Team member** (see `/teams/:slug)

Take user, maybe his submissions (really depends on API response) and render component with user avatar, name, list of submitted exercises and leave button for the current user.

```
<ex-team-member user="{...}"></ex-team-member>
```

These would be great starting points but of course you can choose something else. Other good candidates for directives are editor on submission page, pending submission and submissions filters, notifications etc.

Feel free to let me know if you would like some guidance in the beginning. I can for example create detailed issues for some directives that we'll need, with more detail steps describing the implementation.

Another thing that would be really helpful is to get through the readme in client directory and try to setup client's development pipeline on you local machine. Any feedback on what could be improved in the readme or into which problems did you ran during the setup will be great.

**Comment by purezen** [comment]
Hmm.. Cool!

**Comment by bIgBV** [comment]
@mtakac those ideas are interesting, I'll get started with first establishing a development pipeline and then get working on building some example directives.

**Comment by kytrinyx** [comment]
I'm so sorry that I've failed to follow up on this. The long and short of it is that I don't know how to manage a complex web frontend as open source. There are so many moving pieces. I think we got blocked on a heroku workflow thing.

I'm going to close the issue, but I believe that we've sorted out the heroku buildpack issue, so if anyone wants to take another look at this, you're welcome to do so.
